### DISCUSSION TITLE
vuex unknown action type: Nuxtjs

### DISCUSSION THREAD

**Discussion by mhabbasi-gh** [nuxt/nuxt#9080] [Questions (Nuxt 2)] [answered]
I'm trying to fetch an action in a fetch hook in a page of my nuxt application but I get `[vuex] unknown action type error.` This is my folder setup:

![Screen Shot 2021-03-30 at 7 02 45 PM](https://user-images.githubusercontent.com/58435425/113068559-7fc50780-918c-11eb-93e6-39b0195511d5.png)

This is my `store/index.js`:

```
import vue from "vue";
import Vuex from "vuex";
import axios from "axios";
import Vue from "vue";

import liveEvents from "./liveEvents";

Vue.use(Vuex, axios);

const store = () => {
  return new Vuex.Store({
    modules:{
      liveEvents
    }
  })
}

export default store
```

this is my `store/liveEvents/index.js`:

```
const state = () => ({
  eventsList: []
});

const actions = () => ({
  async eventsList({ commit }) {
   // the actions ...
  },
 
const mutations = () => ( {
  SET_EVENTLIST(state, events) {
    state.eventsList = events;
  },
 
const getters = () => ({
});

export default {
  namespaced: true,
  state,
  actions,
  mutations,
  getters
}  
```
And this is how I call it in my page:
```
async fetch() {
  const { store, error } = this.$nuxt.context;
  try {
    await store.dispatch("liveEvents/eventsList", null, { root: true });
  } catch (e) {
    error({
      message: "error"
    });
  }
}
```
And this is the error I get:
```
[vuex] unknown action type: liveEvents/eventsList 
```
How can I fix this error?

**Comment by clemcode** [comment, marked as answer]
Hi,

in your store modules, only `state` should be a function, keep `getters`, `mutations` and `actions` as plain objects, like so : 

```js
export const state = () => ({
  eventsList: []
})

export const actions = {
  async eventsList({ commit }) {
    // the actions ...
  }
}

export const mutations = {
  SET_EVENTLIST(state, events) {
    state.eventsList = events;
  }
}

// you don't need to export empty objects just for the same of completeness
export const getters = {}

// this is true by default
export const namespaced = true
```

The reason to keep `state` as a function is to avoid shared state on the server as Node is mono-threaded (think of the `data()` option on a traditional Vue component). As `getters`, `mutations` and `actions` are not mutated you should keep them as objects. Exporting a function for `actions` triggered the error you saw.
